[RFC] Generalize pytorch content for non-native device execution

ankurneog opened the pull request:
## Motivation
This PR is for review of the following RFC : Modify PyTorch framework UTs so that non-cuda devices such as intel Gaudi and intel XPU is able to harness the content and improve quality.
https://github.com/pytorch/rfcs/pull/66/files

Commit:
[RFC] Generalize pytorch content for non-native device execution

jgong5 commented on RFC-0039-generalize-pytorch-ut.md:
```diff
@@ -0,0 +1,52 @@
+
+We propose to modify the content wherever possible to make it available for non-cuda device execution
+
+This will also ensure greater participation for content enhancement.
```
I guess it would make sense to elaborate on the problems here before diving into the proposed implementation.

ankurneog replied:
@jgong5  : thanks, have updated with some details.

albanD commented on RFC-0039-generalize-pytorch-ut.md:
```diff
@@ -0,0 +1,52 @@
+
+This will also ensure greater participation for content enhancement.
+
+## **Proposed Implementation**
```
I think the idea makes a lot of sense.
I think we would need more details and how this interracts with existing features like the device-generic tests (https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_device_type.py) that already work for privateuse1 partially btw and the opinfo consistency tests.

Also I don't think we want to aim at running the full test suite with the side device available but select specific device-dependent tests that need to be ran for each device we support.

ankurneog replied:
@albanD  : thanks for your comment , yes I believe we need some extensive hooks to enable this , for eg. the one we introduced with 
https://github.com/pytorch/pytorch/pull/128584/files#diff-d183f2afc51d6a59bc70094e8f476d2468c45e415500f6eb60abad955e065156R531
```@onlyNativeDeviceTypesAnd(["hpu"])```
The other devices can add to to such list, if it supports the TC.

we can modify other hooks like skipIfDevice in similar fashion.

The common_device_type is useable if we replace the ```onlyNativeDeviceType```  decorator. 
It was widely used in the initial files but all recent files are not using it (eg: dynamo/distributed) and instead directly make ```.cuda()``` calls. 

However these content shouldn't be too difficult to migrate. 

I believe in general we should ensure  new test content uses the common_device_type framework, and open up the content for "non-native" device execution.

albanD commented on RFC-0039-generalize-pytorch-ut.md:
```diff
@@ -0,0 +1,52 @@
+*   Dtypes for a device should be dynamically loaded per op based on a common dictionary, instead of using different variables per device , eg: dtypesIfCuda
+* Miscelleneous decorators such as @skipIfCuda should be generalized @skipIfDevice
+*   Extend use of instantiate_device_type for all content, so that developers are forced to use generalized device code rather than using "cuda" or "cpu"
+* Generalize common distributed content , so that it can be extended for non nccl backends such as intel's hccl and ccl
```
cc @wconstab if there is anyone on the distributed side that can help with designing this?

albanD commented on RFC-0039-generalize-pytorch-ut.md:
```diff
@@ -0,0 +1,52 @@
+* Miscelleneous decorators such as @skipIfCuda should be generalized @skipIfDevice
+*   Extend use of instantiate_device_type for all content, so that developers are forced to use generalized device code rather than using "cuda" or "cpu"
+* Generalize common distributed content , so that it can be extended for non nccl backends such as intel's hccl and ccl
+* Generalize the dynamo content for specific backends which other devices might want to verify with existing content.
```
Not sure what this means? But I guess Dynamo tests should happen the same way as other tests (split between device-generic and others and we only re-run the device-generic tests)

ankurneog replied:
@albanD  : yes you are right, what i meant is - for the devices there should be flexibility to execute the TC based backends the device supports ( eg: ```aot_eager```, ```cudagraphs```,```aot_ts```) and also ability to add custom backends which are not part of the tree for eg: intel's ```hpu_backend```

jgong5 commented on RFC-0039-generalize-pytorch-ut.md:
```diff
@@ -0,0 +1,62 @@
+*   Extend use of instantiate_device_type for all content, so that developers are forced to use generalized device code rather than using "cuda" or "cpu"
+*   Generalize common distributed content , so that it can be extended for non nccl backends such as intel's hccl and ccl
+*   Generalize the dynamo content for specific backends which other devices might want to verify with existing content, the backends should always be extracted from
+    a list that is abstracted out and the list can be appended per device per TC.
```
This is a good list of items to generalize the test cases. Does the proposal just focus on the devices having dedicated device tags installed in PyTorch core or also support the PrivateUse1 device which is used to extend PyTorch with any out of the tree devices?

ankurneog replied:
@jgong5  : Thanks for your comment, Since i investigated mostly in lines to support intel Gaudi, which  has dedicated device tag, i have not checked the impact or support needed for PrivateUse1 devices

jgong5 replied:
Perhaps @FFFrog @Yikun have more thoughts on this?

FFFrog replied:
@jgong5 @ankurneog Sorry for the late reply.

In theory, in the test framework of PyTorch, dedicated keys are almost the same as public keys (PrivateUse1), and PrivateUse1 is already supported in the test framework.

First of all, I can`t agree more with this proposal, because Ascend NPU is currently facing the above-described problems; The solution proposed by @ankurneog can solve most of the problems we have encountered; We are currently sorting out all the problems encountered, and will add them to this RFC later, and hope that the new stuff we will add will help the proposal be more complete.

By the way, if possible, we can work together to complete this proposal and make it land in PyTorch :D

ankurneog commented:
Following PRs are submitted addressing some of the issues mentioned in the RFC : 
https://github.com/pytorch/pytorch/pull/128584
https://github.com/pytorch/pytorch/pull/130714
https://github.com/pytorch/pytorch/pull/144387
https://github.com/pytorch/pytorch/pull/145476
https://github.com/pytorch/pytorch/pull/133975
https://github.com/pytorch/pytorch/pull/137863
https://github.com/pytorch/pytorch/pull/131758
https://github.com/pytorch/pytorch/pull/138216
https://github.com/pytorch/pytorch/pull/133209
https://github.com/pytorch/pytorch/pull/134994
https://github.com/pytorch/pytorch/pull/134994
https://github.com/pytorch/pytorch/pull/138216
https://github.com/pytorch/pytorch/pull/144944
https://github.com/pytorch/pytorch/pull/140536
https://github.com/pytorch/pytorch/pull/145056
https://github.com/pytorch/pytorch/pull/139184
https://github.com/pytorch/pytorch/pull/139749
https://github.com/pytorch/pytorch/pull/146472
https://github.com/pytorch/pytorch/pull/140536
https://github.com/pytorch/pytorch/pull/145222
https://github.com/pytorch/pytorch/pull/139184
https://github.com/pytorch/pytorch/pull/139749
https://github.com/pytorch/pytorch/pull/145521
https://github.com/pytorch/pytorch/pull/145589
https://github.com/pytorch/pytorch/pull/146290
https://github.com/pytorch/pytorch/pull/146478
